### ISSUE TITLE
Unable to establish traffic controller websocket through proxy on custom domain

### ISSUE THREAD

**Issue by amaltson** [cloudfoundry-incubator/lattice-release#223]
I tried launching a droplet while it throw the below error, the droplet actually started anyway.

```
$ ltc ld app droplet
No port specified. Defaulting to 8080.
Creating App: app
Error dialing traffic controller server: Tunnel Forbidden.
Please ask your Cloud Foundry Operator to check the platform configuration (traffic controller is ws://doppler.cf.otpp.me).
.Error dialing traffic controller server: Tunnel Forbidden.
Please ask your Cloud Foundry Operator to check the platform configuration (traffic controller is ws://doppler.cf.otpp.me).
Error dialing traffic controller server: Tunnel Forbidden.
Please ask your Cloud Foundry Operator to check the platform configuration (traffic controller is ws://doppler.cf.otpp.me).
.Error dialing traffic controller server: Tunnel Forbidden.
Please ask your Cloud Foundry Operator to check the platform configuration (traffic controller is ws://doppler.cf.otpp.me).
Error dialing traffic controller server: Tunnel Forbidden.
Please ask your Cloud Foundry Operator to check the platform configuration (traffic controller is ws://doppler.cf.otpp.me).
.panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x20 pc=0x519fa8]

goroutine 22 [running]:
github.com/cloudfoundry-incubator/lattice/ltc/logs/console_tailed_logs_outputter.(*ConsoleTailedLogsOutputter).errorCallback(0xc2080b1050, 0x0, 0x0)
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/logs/console_tailed_logs_outputter/console_tailed_logs_outputter.go:66 +0x28
github.com/cloudfoundry-incubator/lattice/ltc/logs/console_tailed_logs_outputter.*ConsoleTailedLogsOutputter.(github.com/cloudfoundry-incubator/lattice/ltc/logs/console_tailed_logs_outputter.errorCallback)·fm(0x0, 0x0)
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/logs/console_tailed_logs_outputter/console_tailed_logs_outputter.go:48 +0x3b
github.com/cloudfoundry-incubator/lattice/ltc/logs.(*logReader).readChannels(0xc2080bce40, 0xc208054a80, 0xc208054c00, 0xc20814a080, 0xc20814a090)
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/logs/logs.go:47 +0x108
github.com/cloudfoundry-incubator/lattice/ltc/logs.(*logReader).TailLogs(0xc2080bce40, 0x7ffd117d1489, 0x17, 0xc20814a080, 0xc20814a090)
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/logs/logs.go:33 +0x108
created by github.com/cloudfoundry-incubator/lattice/ltc/logs/console_tailed_logs_outputter.(*ConsoleTailedLogsOutputter).OutputTailedLogs
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/logs/console_tailed_logs_outputter/console_tailed_logs_outputter.go:48 +0x122

goroutine 1 [chan receive]:
github.com/pivotal-golang/clock.(*realClock).Sleep(0x114ee10, 0x3b9aca00)
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/lattice/ltc/Godeps/_workspace/src/github.com/pivotal-golang/clock/clock.go:24 +0x73
github.com/cloudfoundry-incubator/lattice/ltc/app_runner/command_factory.(*AppRunnerCommandFactory).pollUntilSuccess(0xc2080b4fc0, 0x1bf08eb000, 0xc2081c10f8, 0x1, 0x0)
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/app_runner/command_factory/app_runner_command_factory.go:424 +0x231
github.com/cloudfoundry-incubator/lattice/ltc/app_runner/command_factory.(*AppRunnerCommandFactory).pollUntilAllInstancesRunning(0xc2080b4fc0, 0x1bf08eb000, 0x7ffd117d1489, 0x17, 0x1, 0xb87be0, 0x5, 0x0)
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/app_runner/command_factory/app_runner_command_factory.go:391 +0x91
github.com/cloudfoundry-incubator/lattice/ltc/app_runner/command_factory.(*AppRunnerCommandFactory).WaitForAppCreation(0xc2080b4fc0, 0x7ffd117d1489, 0x17, 0x1bf08eb000, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/app_runner/command_factory/app_runner_command_factory.go:335 +0x1bc
github.com/cloudfoundry-incubator/lattice/ltc/droplet_runner/command_factory.(*DropletRunnerCommandFactory).launchDroplet(0xc2080b4fc0, 0xc20805c000)
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/droplet_runner/command_factory/droplet_runner_command_factory.go:578 +0x14a6
github.com/cloudfoundry-incubator/lattice/ltc/droplet_runner/command_factory.*DropletRunnerCommandFactory.(github.com/cloudfoundry-incubator/lattice/ltc/droplet_runner/command_factory.launchDroplet)·fm(0xc20805c000)
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/droplet_runner/command_factory/droplet_runner_command_factory.go:238 +0x31
github.com/codegangsta/cli.Command.Run(0xbaf150, 0xe, 0x0, 0x0, 0xc20802b7a0, 0x1, 0x1, 0xc5de50, 0x2f, 0xd1ab50, ...)
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/lattice/ltc/Godeps/_workspace/src/github.com/codegangsta/cli/command.go:127 +0xffe
github.com/codegangsta/cli.(*App).Run(0xc208066300, 0xc20800a000, 0x4, 0x4, 0x0, 0x0)
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/lattice/ltc/Godeps/_workspace/src/github.com/codegangsta/cli/app.go:159 +0xd7b
main.main()
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/main.go:10 +0x50

goroutine 5 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
	/usr/local/go/src/os/signal/signal_unix.go:27 +0x35

goroutine 6 [chan receive]:
github.com/cloudfoundry-incubator/lattice/ltc/exit_handler.(*exitHandler).Run(0xc20802eb40)
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/exit_handler/exit_handler.go:36 +0x4f
created by github.com/cloudfoundry-incubator/lattice/ltc/setup_cli.NewCliApp
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/setup_cli/setup_cli.go:29 +0x550

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1

goroutine 18 [IO wait]:
net.(*pollDesc).Wait(0xc208136060, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc208136060, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc208136000, 0xc20800f000, 0x1000, 0x1000, 0x0, 0x7f469c02ed48, 0xc208134010)
	/usr/local/go/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc208146000, 0xc20800f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:121 +0xdc
net/http.noteEOFReader.Read(0x7f469c031828, 0xc208146000, 0xc208074b58, 0xc20800f000, 0x1000, 0x1000, 0xa08ca0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1270 +0x6e
net/http.(*noteEOFReader).Read(0xc208142040, 0xc20800f000, 0x1000, 0x1000, 0xc208012000, 0x0, 0x0)
	<autogenerated>:125 +0xd4
bufio.(*Reader).fill(0xc20814c000)
	/usr/local/go/src/bufio/bufio.go:97 +0x1ce
bufio.(*Reader).Peek(0xc20814c000, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xf0
net/http.(*persistConn).readLoop(0xc208074b00)
	/usr/local/go/src/net/http/transport.go:842 +0xa4
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:660 +0xc9f

goroutine 19 [select]:
net/http.(*persistConn).writeLoop(0xc208074b00)
	/usr/local/go/src/net/http/transport.go:945 +0x41d
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:661 +0xcbc

goroutine 41 [select]:
net/http.(*persistConn).writeLoop(0xc208074bb0)
	/usr/local/go/src/net/http/transport.go:945 +0x41d
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:661 +0xcbc

goroutine 40 [IO wait]:
net.(*pollDesc).Wait(0xc20808c060, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc20808c060, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc20808c000, 0xc20814f000, 0x1000, 0x1000, 0x0, 0x7f469c02ed48, 0xc20805eeb8)
	/usr/local/go/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc208170000, 0xc20814f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:121 +0xdc
net/http.noteEOFReader.Read(0x7f469c031828, 0xc208170000, 0xc208074c08, 0xc20814f000, 0x1000, 0x1000, 0xa08ca0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1270 +0x6e
net/http.(*noteEOFReader).Read(0xc2081320e0, 0xc20814f000, 0x1000, 0x1000, 0xc208013b00, 0x0, 0x0)
	<autogenerated>:125 +0xd4
bufio.(*Reader).fill(0xc20814c0c0)
	/usr/local/go/src/bufio/bufio.go:97 +0x1ce
bufio.(*Reader).Peek(0xc20814c0c0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xf0
net/http.(*persistConn).readLoop(0xc208074bb0)
	/usr/local/go/src/net/http/transport.go:842 +0xa4
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:660 +0xc9f

goroutine 12 [chan receive]:
github.com/cloudfoundry-incubator/lattice/ltc/logs/console_tailed_logs_outputter.(*ConsoleTailedLogsOutputter).OutputTailedLogs(0xc2080b1050, 0x7ffd117d1489, 0x17)
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/logs/console_tailed_logs_outputter/console_tailed_logs_outputter.go:50 +0x16d
created by github.com/cloudfoundry-incubator/lattice/ltc/app_runner/command_factory.(*AppRunnerCommandFactory).WaitForAppCreation
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/go/src/github.com/cloudfoundry-incubator/lattice/ltc/app_runner/command_factory/app_runner_command_factory.go:332 +0x119

goroutine 25 [chan receive]:
github.com/cloudfoundry/noaa.func·003()
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/lattice/ltc/Godeps/_workspace/src/github.com/cloudfoundry/noaa/consumer.go:86 +0x87
created by github.com/cloudfoundry/noaa.(*Consumer).TailingLogsWithoutReconnect
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/lattice/ltc/Godeps/_workspace/src/github.com/cloudfoundry/noaa/consumer.go:91 +0x400

goroutine 14 [chan receive]:
github.com/cloudfoundry/noaa.func·003()
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/lattice/ltc/Godeps/_workspace/src/github.com/cloudfoundry/noaa/consumer.go:86 +0x87
created by github.com/cloudfoundry/noaa.(*Consumer).TailingLogsWithoutReconnect
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/lattice/ltc/Godeps/_workspace/src/github.com/cloudfoundry/noaa/consumer.go:91 +0x400

goroutine 27 [chan receive]:
github.com/cloudfoundry/noaa.func·003()
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/lattice/ltc/Godeps/_workspace/src/github.com/cloudfoundry/noaa/consumer.go:86 +0x87
created by github.com/cloudfoundry/noaa.(*Consumer).TailingLogsWithoutReconnect
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/lattice/ltc/Godeps/_workspace/src/github.com/cloudfoundry/noaa/consumer.go:91 +0x400

goroutine 16 [chan receive]:
github.com/cloudfoundry/noaa.func·003()
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/lattice/ltc/Godeps/_workspace/src/github.com/cloudfoundry/noaa/consumer.go:86 +0x87
created by github.com/cloudfoundry/noaa.(*Consumer).TailingLogsWithoutReconnect
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/lattice/ltc/Godeps/_workspace/src/github.com/cloudfoundry/noaa/consumer.go:91 +0x400

goroutine 51 [chan receive]:
github.com/cloudfoundry/noaa.func·003()
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/lattice/ltc/Godeps/_workspace/src/github.com/cloudfoundry/noaa/consumer.go:86 +0x87
created by github.com/cloudfoundry/noaa.(*Consumer).TailingLogsWithoutReconnect
	/tmp/build/b86c715a-5e20-4cc9-5a5f-429840fd1200/lattice/ltc/Godeps/_workspace/src/github.com/cloudfoundry/noaa/consumer.go:91 +0x400
```

**Comment by davidwadden** [comment]
That's from trying to stream logs back from loggregator during app startup.  I've seen it intermittently, often due to OS X DNS resolution of `xip.io` failing.

But, on closer inspection, your error looks slightly different:

```
Error dialing traffic controller server: Tunnel Forbidden.
Please ask your Cloud Foundry Operator to check the platform configuration (traffic controller is ws://doppler.cf.otpp.me).
```

We'll need to look into where it's coming from.  Not sure why it's not letting you open a websocket to pull logs.  Anything you can offer about your networking configuration?

**Comment by amaltson** [comment]
Hey @davidwadden, sure thing. That particular message is happening when I tried to launch a droplet from a different subnet in a datacenter, whereas Lattice is on a developer machine in the same building as me.

The DNS is an external DNS that points to internal private IP addresses.

The Lattice version is 0.4.3 for client/server, I haven't had a chance to upgrade yet.

I'm able to deploy from all the subnets in the same building, but not outside. It could be a networking issue, so I'll go look into it. Other Lattice commands seem to work, although I was getting an issue with `ltc rm` where I was getting a 415 but can't reproduce it again.

Thank you, I'll report back.

**Comment by amaltson** [comment]
Hey @davidwadden, I tested to see if WebSocket connections were coming through our network. Using the [faye-websocket-ruby gem](https://github.com/faye/faye-websocket-ruby), I set up a server and connected a client to it. Here's the [gist of the client and server code](https://gist.github.com/amaltson/5fadbca01c39de777cc5)

```
build-server $ thin start
Using rack adapter
Thin web server (v1.6.4 codename Gob Bluth)
Maximum connections set to 1024
Listening on 0.0.0.0:3000, CTRL+C to stop

hephaestus $ ruby websocket-client.rb build-server
[:open]
[:message, "Hello, world!"]
```

I've also tried the other way around, having the server run locally and having `build-server` connect to my machine. So it's not looking to be an issue with websockets... any idea what else it could be? Thanks again!

**Comment by ekcasey** [comment]
We are having a hard time reproducing this error. Can you tell us more about your lattice deployment?

How did you deploy your lattice cluster (i.e. vagrant, terraform, or other)?
Are you connecting via proxy?

We were unable to find any reference to "Tunnel Forbidden" in either Golang's or our own source code, and it's not a standard HTTP exit code semantic value. Could it be a custom response from your proxy?

**Comment by amaltson** [comment]
@ekcasey sorry for the trouble! This Lattice is deployed using the manual setup instructions.

Your question about the proxy led me down the exact issue. I tried this on another server in the same subnet but one that doesn't use the http proxy environment variables and logs are streaming fine. We use an Apache Traffic server to give some servers internet access, clearly it's not configured for WebSockets :wink:. I just need to add some more entries to the `no_proxy`.

Thanks again for looking into this, sorry for the trouble!

**Comment by mbechauf** [comment]
I have the same issue. @amaltson How did you end up fixing this? You said you added more entries to `no_proxy`, but I'm not clear what. I'm just getting the same weird `traffic controller is ws://doppler.` message, but I can't make sense of it.

**Comment by davidwadden** [comment]
that's interesting.  

@mbechauf - if you're using our bundles (and not doing a manual install like @amaltson above), we set the `no_proxy` using the following at [`Vagrantfile:16-21`](https://github.com/cloudfoundry-incubator/lattice-release/blob/master/vagrant/Vagrantfile#L16-L21).

```
    config.proxy.no_proxy = [
      "localhost", "127.0.0.1",
      (ENV["LATTICE_IP"] || "192.168.11.11"),
      (ENV["LATTICE_DOMAIN"] || "192.168.11.11.xip.io"),
      ".consul"
    ].join(',')
``` 

@amaltson - can you give some more detail on your general configuration?

you said it was manual install, so i'm to assume there was no vagrant?  
did you set the proxy for the cluster or just for `ltc`'s context?

we have two places that support proxies in lattice - one is the cluster proxy (`http_proxy` and `https_proxy` env vars at `vagrant up`), the other is the CLI proxy (env vars when using `ltc`), which did you need to set no_proxy for?

i'm going to reopen this to discuss better support for proxies in custom domains

**Comment by mbechauf** [comment]
Wanted to use the opportunity to discuss my custom domain setup for Windows. First of all, I think every setup behind firewall necessitates a custom domain. I tried to work with the `xip.io` domains, and it doesn't work. The reason is that names like `myapp.myip.xip.io` cannot be resolved if you are behind a firewall, at least not in the firewall setup that SAP has in place. DNS names can only be resolved if they are in the SAP domain, not in the case of public Internet names. 

What's worse is that even http requests to an address like `myapp.myip.xip.io` cannot work because while the corporate proxy is likely able to resolve correctly to `myip`, it then of course assumes that `myip` is a public Internet address, and not a local address on the requester's workstation.

I had the same problems starting with `0.6.0` also with `local.lattice.cf`.

So, I'm surprised that in the documentation about Lattice proxy setup there is no hint that any proxy usually makes a DNSMasq-like setup necessary. 

Now, on to Windows. Unfortunately, there is no DNSMasq available for Windows, so what I ended up doing was setting up a PowerDNS instance on my Windows machine on an Ubuntu Virtualbox image, and installing the `pipe` backend together with the XIP PowerDNS code available at https://github.com/basecamp/xip-pdns. I then added the IP of the PowerDNS Virtualbox as a custom DNS server to my Windows IP, and that did the trick. 

So, this is exactly the setup I am using now. Since `0.6.0` now defaults to `local.lattice.cf` this also means that I have to use `LATTICE_DOMAIN` and set it to `192.168.11.11.xip.io` where `xip.io` is actually resolved now locally by my PowerDNS instance running on my machine. As a result `ping 192.168.11.11.xip.io` actually correctly works and resolves to `192.168.11.11` because the `xip.io` domain is resolved by my local DNS. 

As a result, starting with `0.6.0` my assumption is that every proxy setup necessitates a custom domain setup.
